script/script.go

```diff
@@ -25,18 +25,86 @@ import (
 	gjson "layeh.com/gopher-json"
 )
 
-// InitialPoolSize is the number of lua states created initially per route
-var InitialPoolSize int = 3
+const (
+	defaultPoolSize int = 1000
+	genericStateKey     = "generic"
+)
+
+func newState() (*lua.LState, error) {
+	L := lua.NewState()
+	L.PreloadModule("base64", base64.Loader)
+	L.PreloadModule("http", gluahttp.NewHttpModule(&http.Client{}).Loader)
+	L.PreloadModule("url", gluaurl.Loader)
+	L.PreloadModule("json", gjson.Loader)
+	L.SetGlobal("print", L.NewFunction(printToLog))
+	L.SetGlobal("sleep", L.NewFunction(sleep))
+	return L, nil
+}
+
+type statePool struct {
+	pool chan *lua.LState
+}
+
+func newStatePool(max int) (*statePool, error) {
+	sp := &statePool{
+		pool: make(chan *lua.LState, max),
+	}
+	for i := 0; i < max; i++ {
+		state, err := newState()
+		if err != nil {
+			return nil, err
+		}
+		sp.pool <- state
+	}
```

Comment:
I think we may start with empty pool (or configurable min size) and allow it to grow organically. Assuming max size is passed as a flag this would allow us to set it to some quite high number while real amount of utilized states would be defined by traffic rate shape.

Reply:
The drawback would be that this could lead to latency spikes, because we run and rely on autoscaling.
Assumption is that this can lead to a latency spike (not 100% sure but in tests it felt a bit like this).